client/src/components/general/VirtualizedList.tsx

```diff
@@ -0,0 +1,80 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import styles from './VirtualizedList.module.scss';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
+import { ListItem } from "../../models/ListItem";
+
+type Props = {
+    items: ListItem[],
+    itemHeight: number
+}
+
+export function VirtualizedList(props: Props) {  
+  const { items, itemHeight } = props;
+  const [selectedItem, setSelectedItem] = useState<number | null>(null);
+  const [visibleItems, setVisibleItems] = useState<ListItem[]>([]);
```

Comment:
Storing the entire slice of visible items in the state as a list of items is inefficient. This is because each scroll will trigger this useEffect - and every time this effect runs it's storing the slice in the state and that will cause a re-render of the component (because it's a new instance of an array), even if the visible items don't actually change (when a scroll is too small to pass the threshold of an item's height). This can lead to performance bottlenecks, especially with large lists or rapid scrolling.
I would suggest storing the firstIndex in the state and calculating the lastIndex based on the containerHeight at render time (outside a useEffect, before the return)

Reply:
you're right. I tried saving startIndex to state and calculate endIndex at render time but encountered a different issue: In the first render the container ref was still null so endIndex was 1 (and only one item was displayed on screen), and when useEffect was triggered and called setStartIndex it still didn't cause a re-render since the index stayed 0, until I triggered a re-render by scrolling or searching text and everything was calculated well. So instead I tried saving endIndex to state and calculate startIndex at render time with containerHeight, that way the useEffect was triggered, changed the endIndex and caused a re-render where container ref was not null and endIndex was calculated accordingly. Do you think this solution is Ok? Do you have another suggestion? Thanks :)